File: server/routes/jobs.js
const express = require('express');
const { v4: uuidv4 } = require('uuid');
const db = require('../config/db');
const { authenticate, authorize } = require('../middleware/auth');

const router = express.Router();

// POST /api/jobs/:id/quick-order — Direct order a job (skips bidding, creates a transaction)
router.post('/:id/quick-order', authenticate, async (req, res) => {
  try {
    const job = await db.prepare('SELECT * FROM jobs WHERE id = ? AND status = ?').get(req.params.id, 'open');
    if (!job) return res.status(404).json({ error: 'Job not found or not open for orders' });

    const { amount, proposal } = req.body;
    if (!amount || parseFloat(amount) <= 0) {
      return res.status(400).json({ error: 'A valid amount is required' });
    }

    // Update job status — award to this user
    await db.prepare('UPDATE jobs SET status = ?, awarded_to = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ?').run('in_progress', req.user.id, req.params.id);

    // Create transaction in escrow (following the same pattern as the award endpoint)
    const txnId = uuidv4();
    await db.prepare('INSERT INTO transactions (id, job_id, client_id, freelancer_id, amount, status) VALUES (?, ?, ?, ?, ?, ?)')
      .run(txnId, req.params.id, job.client_id, req.user.id, parseFloat(amount), 'escrow');

    // Create notification for the job poster
    const notifId = uuidv4();
    await db.prepare('INSERT INTO notifications (id, user_id, title, message, type) VALUES (?, ?, ?, ?, ?)')
      .run(notifId, job.client_id, 'New Order on Your Job!',
        `${req.user.full_name || 'A freelancer'} placed a quick order on your job "${job.title}" for ETB ${parseFloat(amount).toLocaleString()}.`,
        'order');

    res.status(201).json({ transactionId: txnId, message: 'Quick order placed successfully!' });
  } catch (err) {
    console.error('Error placing quick order:', err);
    res.status(500).json({ error: 'Failed to place quick order' });
  }
});

// GET /api/jobs - List jobs
router.get('/', async (req, res) => {
  try {
    const { category, search, minBudget, maxBudget, sort, status, clientId, freelancerId } = req.query;
    let query = `
      SELECT j.*, u.full_name as client_name, u.profile_picture as client_picture,
             u.rating as client_rating
      FROM jobs j
      JOIN users u ON j.client_id = u.id
      WHERE 1=1
    `;
    const params = [];

    if (category) { query += ' AND j.category = ?'; params.push(category); }
    if (search) { query += ' AND (j.title LIKE ? OR j.description LIKE ?)'; params.push(`%${search}%`, `%${search}%`); }
    if (minBudget) { query += ' AND j.budget_max >= ?'; params.push(parseFloat(minBudget)); }
    if (maxBudget) { query += ' AND j.budget_min <= ?'; params.push(parseFloat(maxBudget)); }
    if (status) { query += ' AND j.status = ?'; params.push(status); }
    if (clientId) { query += ' AND j.client_id = ?'; params.push(clientId); }
    if (freelancerId) { query += ' AND j.awarded_to = ?'; params.push(freelancerId); }

    switch (sort) {
      case 'budget_high': query += ' ORDER BY j.budget_max DESC'; break;
      case 'budget_low': query += ' ORDER BY j.budget_min ASC'; break;
      case 'newest': query += ' ORDER BY j.created_at DESC'; break;
      default: query += ' ORDER BY j.created_at DESC';
    }

    const jobs = await db.prepare(query).all(...params);

    // Get bid counts and deliveries for each job
    const jobsWithMeta = [];
    for (const j of jobs) {
      const bidCount = await db.prepare('SELECT COUNT(*) as count FROM bids WHERE job_id = ?').get(j.id).count;
      const delivery = await db.prepare('SELECT * FROM job_deliveries WHERE job_id = ? ORDER BY created_at DESC LIMIT 1').get(j.id);
      let deliveryFiles = [];
      try { if (delivery) deliveryFiles = JSON.parse(delivery.files || '[]'); } catch {}
      jobsWithMeta.push({ 
        ...j, 
        bid_count: bidCount,
        has_delivery: !!delivery,
        delivery_file_count: deliveryFiles.length,
        delivery_summary: delivery ? { 
          id: delivery.id, 
          message: delivery?.message, 
          file_count: deliveryFiles.length,
          files: deliveryFiles.slice(0, 3),
          created_at: delivery.created_at 
        } : null 
      });
    }

    res.json({ jobs: jobsWithMeta });
  } catch (err) {
    console.error('Error listing jobs:', err);
    res.status(500).json({ error: 'Failed to list jobs' });
  }
});

// GET /api/jobs/:id
router.get('/:id', async (req, res) => {
  try {
    const job = await db.prepare(`
      SELECT j.*, 
        u.full_name as client_name, u.profile_picture as client_picture,
        u.rating as client_rating, u.phone as client_phone, u.city as client_city,
        af.full_name as awarded_name, af.profile_picture as awarded_picture,
        af.rating as awarded_rating, af.verified as awarded_verified, af.city as awarded_city
      FROM jobs j 
      JOIN users u ON j.client_id = u.id 
      LEFT JOIN users af ON j.awarded_to = af.id
      WHERE j.id = ?
    `).get(req.params.id);

    if (!job) return res.status(404).json({ error: 'Job not found' });

    const bids = await db.prepare(`
      SELECT b.*, u.full_name as freelancer_name, u.profile_picture as freelancer_picture,
             u.rating as freelancer_rating, u.verified as freelancer_verified, u.city as freelancer_city
      FROM bids b JOIN users u ON b.freelancer_id = u.id WHERE b.job_id = ? ORDER BY b.amount ASC
    `).all(job.id);

    // Also fetch any active transactions for this job (for quick order info)
    const transactions = await db.prepare(`
      SELECT t.*, 
        u.full_name as freelancer_name, u.profile_picture as freelancer_picture,
        u.rating as freelancer_rating
      FROM transactions t 
      LEFT JOIN users u ON t.freelancer_id = u.id 
      WHERE t.job_id = ?
      ORDER BY t.created_at DESC
    `).all(job.id);

    // Fetch deliveries for this job
    const deliveries = await db.prepare('SELECT * FROM job_deliveries WHERE job_id = ? ORDER BY created_at DESC').all(job.id);

    res.json({ job, bids, transactions, deliveries });
  } catch (err) {
    console.error('Error fetching job:', err);
    res.status(500).json({ error: 'Failed to fetch job' });
  }
});

// POST /api/jobs - Create a job
router.post('/', authenticate, authorize('client'), async (req, res) => {
  try {
    const { title, description, budgetMin, budgetMax, category, deadline } = req.body;
    if (!title || !description || !budgetMin || !budgetMax || !category) {
      return res.status(400).json({ error: 'Title, description, budget range, and category are required' });
    }

    const id = uuidv4();
    await db.prepare(`
      INSERT INTO jobs (id, client_id, title, description, budget_min, budget_max, category, deadline)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?)
    `).run(id, req.user.id, title, description, parseFloat(budgetMin), parseFloat(budgetMax), category, deadline || null);

    const job = await db.prepare('SELECT * FROM jobs WHERE id = ?').get(id);
    res.status(201).json({ job });
  } catch (err) {
    console.error('Error creating job:', err);
    res.status(500).json({ error: 'Failed to create job' });
  }
});

// POST /api/jobs/:id/bid - Place a bid on a job
router.post('/:id/bid', authenticate, authorize('freelancer'), async (req, res) => {
  try {
    const job = await db.prepare('SELECT * FROM jobs WHERE id = ? AND status = ?').get(req.params.id, 'open');
    if (!job) return res.status(404).json({ error: 'Job not found or not accepting bids' });

    const { amount, proposal } = req.body;
    if (!amount) return res.status(400).json({ error: 'Bid amount is required' });

    const existingBid = await db.prepare('SELECT id FROM bids WHERE job_id = ? AND freelancer_id = ?').get(req.params.id, req.user.id);
    if (existingBid) return res.status(409).json({ error: 'You already placed a bid on this job' });

    const id = uuidv4();
    await db.prepare('INSERT INTO bids (id, job_id, freelancer_id, amount, proposal) VALUES (?, ?, ?, ?, ?)')
      .run(id, req.params.id, req.user.id, parseFloat(amount), proposal || null);

    // Create notification for client
    const notifId = uuidv4();
    await db.prepare('INSERT INTO notifications (id, user_id, title, message, type) VALUES (?, ?, ?, ?, ?)')
      .run(notifId, job.client_id, 'New Bid Received', `${req.user.full_name} placed a bid of ETB ${amount} on your job "${job.title}"`, 'bid');

    const bid = await db.prepare('SELECT * FROM bids WHERE id = ?').get(id);
    res.status(201).json({ bid });
  } catch (err) {
    console.error('Error placing bid:', err);
    res.status(500).json({ error: 'Failed to place bid' });
  }
});

// PUT /api/jobs/:id/award - Award job to a freelancer
router.put('/:id/award', authenticate, authorize('client'), async (req, res) => {
  try {
    const { freelancerId } = req.body;
    const job = await db.prepare('SELECT * FROM jobs WHERE id = ? AND client_id = ? AND status = ?').get(req.params.id, req.user.id, 'open');
    if (!job) return res.status(404).json({ error: 'Job not found or not open' });

    const bid = await db.prepare('SELECT * FROM bids WHERE job_id = ? AND freelancer_id = ? AND status = ?').get(req.params.id, freelancerId, 'pending');
    if (!bid) return res.status(404).json({ error: 'Bid not found' });

    await db.prepare('UPDATE jobs SET status = ?, awarded_to = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ?').run('in_progress', freelancerId, req.params.id);
    await db.prepare('UPDATE bids SET status = ? WHERE job_id = ? AND freelancer_id = ?').run('accepted', req.params.id, freelancerId);
    await db.prepare('UPDATE bids SET status = ? WHERE job_id = ? AND freelancer_id != ? AND status = ?').run('rejected', req.params.id, freelancerId, 'pending');

    // Create transaction in escrow
    const txnId = uuidv4();
    await db.prepare('INSERT INTO transactions (id, job_id, client_id, freelancer_id, amount, status) VALUES (?, ?, ?, ?, ?, ?)')
      .run(txnId, req.params.id, req.user.id, freelancerId, bid.amount, 'escrow');

    // Notify freelancer
    const notifId = uuidv4();
    await db.prepare('INSERT INTO notifications (id, user_id, title, message, type) VALUES (?, ?, ?, ?, ?)')
      .run(notifId, freelancerId, 'Job Awarded!', `You have been awarded the job "${job.title}"!`, 'award');

    res.json({ message: 'Job awarded successfully', transactionId: txnId });
  } catch (err) {
    console.error('Error awarding job:', err);
    res.status(500).json({ error: 'Failed to award job' });
  }
});

// DELETE /api/jobs/:id - Delete a job (only the owner can delete)
router.delete('/:id', authenticate, async (req, res) => {
  try {
    const job = await db.prepare('SELECT * FROM jobs WHERE id = ? AND client_id = ?').get(req.params.id, req.user.id);
    if (!job) {
      return res.status(404).json({ error: 'Job not found or unauthorized' });
    }
    // Soft delete — set status to cancelled
    await db.prepare("UPDATE jobs SET status = 'cancelled', updated_at = CURRENT_TIMESTAMP WHERE id = ?").run(req.params.id);
    res.json({ message: 'Job cancelled successfully' });
  } catch (err) {
    console.error('Error deleting job:', err);
    res.status(500).json({ error: 'Failed to delete job' });
  }
});

// PUT /api/jobs/:id/deliver - Freelancer delivers finished work with files
router.put('/:id/deliver', authenticate, async (req, res) => {
  try {
    const job = await db.prepare("SELECT * FROM jobs WHERE id = ? AND status = 'in_progress'").get(req.params.id);
    if (!job) return res.status(404).json({ error: 'Job not found or not in progress' });
    if (job.awarded_to !== req.user.id) return res.status(403).json({ error: 'Only the awarded freelancer can deliver' });

    const { message, files } = req.body;

    if (files && Array.isArray(files) && files.length > 0) {
      if (files.length > 10) return res.status(400).json({ error: 'Maximum 10 files per delivery' });

      for (const f of files) {
        if (!f.name || !f.type || !f.data)
          return res.status(400).json({ error: 'Each file must have name, type, and data fields' });

        const estimatedBytes = (f.data?.length || 0) * 0.75;
        if (estimatedBytes > 100 * 1024 * 1024) {
          return res.status(400).json({ error: `File "${f.name}" exceeds 100MB limit` });
        }
      }

      const deliveryId = uuidv4();
      await db.prepare(`
        INSERT INTO job_deliveries (id, job_id, freelancer_id, message, files)
        VALUES ($1, $2, $3, $4, $5)
      `).run(deliveryId, req.params.id, req.user.id, message || '', JSON.stringify(files));
      console.log(`📁 Job delivery saved for job ${req.params.id.slice(0, 8)}: ${files.length} file(s)`);
    }

    await db.prepare("UPDATE jobs SET status = 'delivered', updated_at = CURRENT_TIMESTAMP WHERE id = ?").run(req.params.id);

    // Notify client
    const notifId = uuidv4();
    await db.prepare('INSERT INTO notifications (id, user_id, title, message, type) VALUES (?, ?, ?, ?, ?)')
      .run(notifId, job.client_id, 'Work Delivered!',
        `${req.user.full_name || 'The freelancer'} has delivered the finished work for "${job.title}". Please review and approve.`,
        'order');

    res.json({ message: 'Work delivered successfully', deliverySaved: !!(files && files.length > 0) });
  } catch (err) {
    console.error('Error delivering job work:', err);
    res.status(500).json({ error: 'Failed to deliver work' });
  }
});

// PUT /api/jobs/:id/status - Update job status (complete, cancel)
router.put('/:id/status', authenticate, async (req, res) => {
  try {
    const { status } = req.body;
    const job = await db.prepare('SELECT * FROM jobs WHERE id = ?').get(req.params.id);
    if (!job) return res.status(404).json({ error: 'Job not found' });

    if (status === 'completed') {
      if (job.client_id !== req.user.id) return res.status(403).json({ error: 'Only the client can mark as completed' });
    }

    await db.prepare('UPDATE jobs SET status = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ?').run(status, req.params.id);
    res.json({ message: 'Job status updated' });
  } catch (err) {
    console.error('Error updating job status:', err);
    res.status(500).json({ error: 'Failed to update job status' });
  }
});

module.exports = router;

File: python_backend/routers/jobs.py
"""Job routes: CRUD, bids, award, deliver, quick order, status updates."""

import uuid
import json
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from python_backend.database import get_db, ilike_sql
from python_backend.auth import get_current_user
from python_backend.cache import cache_ttl
from python_backend.schemas import (
    JobCreateRequest, BidRequest, AwardRequest, QuickOrderRequest,
    DeliverRequest, JobStatusUpdate,
)

router = APIRouter(prefix="/api/jobs", tags=["jobs"])


@router.get("")
@cache_ttl(seconds=20)
async def list_jobs(
    category: str = None,
    search: str = None,
    minBudget: float = None,
    maxBudget: float = None,
    sort: str = None,
    status: str = None,
    clientId: str = None,
    freelancerId: str = None,
    db: AsyncSession = Depends(get_db),
):
    query = """
        SELECT j.*, u.full_name as client_name, u.profile_picture as client_picture,
               u.rating as client_rating
        FROM jobs j
        JOIN users u ON j.client_id = u.id
        WHERE 1=1
    """
    params = {}

    if category:
        query += " AND j.category = :cat"
        params["cat"] = category
    if search:
        query += f" AND (j.title {ilike_sql()} :search OR j.description {ilike_sql()} :search2)"
        params["search"] = f"%{search}%"
        params["search2"] = f"%{search}%"
    if minBudget is not None:
        query += " AND j.budget_max >= :min_b"
        params["min_b"] = minBudget
    if maxBudget is not None:
        query += " AND j.budget_min <= :max_b"
        params["max_b"] = maxBudget
    if status:
        query += " AND j.status = :stat"
        params["stat"] = status
    if clientId:
        query += " AND j.client_id = :cid"
        params["cid"] = clientId
    if freelancerId:
        query += " AND j.awarded_to = :fid"
        params["fid"] = freelancerId

    sort_map = {"budget_high": "j.budget_max DESC", "budget_low": "j.budget_min ASC", "newest": "j.created_at DESC"}
    query += f" ORDER BY {sort_map.get(sort, 'j.created_at DESC')}"

    result = await db.execute(text(query), params)
    jobs = result.mappings().all()

    jobs_with_meta = []
    for j in jobs:
        bid_count_result = await db.execute(
            text("SELECT COUNT(*) as count FROM bids WHERE job_id = :jid"), {"jid": j["id"]}
        )
        bid_count = bid_count_result.mappings().first()["count"]
        jobs_with_meta.append({**dict(j), "bid_count": bid_count})

    return {"jobs": jobs_with_meta}


@router.get("/{job_id}")
async def get_job(job_id: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(
        text("""
            SELECT j.*,
                u.full_name as client_name, u.profile_picture as client_picture,
                u.rating as client_rating, u.phone as client_phone, u.city as client_city,
                af.full_name as awarded_name, af.profile_picture as awarded_picture,
                af.rating as awarded_rating, af.verified as awarded_verified, af.city as awarded_city
            FROM jobs j
            JOIN users u ON j.client_id = u.id
            LEFT JOIN users af ON j.awarded_to = af.id
            WHERE j.id = :id
        """),
        {"id": job_id},
    )
    job = result.mappings().first()
    if not job:
        raise HTTPException(404, "Job not found")

    bids_result = await db.execute(
        text("""
            SELECT b.*, u.full_name as freelancer_name, u.profile_picture as freelancer_picture,
                   u.rating as freelancer_rating, u.verified as freelancer_verified, u.city as freelancer_city
            FROM bids b JOIN users u ON b.freelancer_id = u.id WHERE b.job_id = :jid ORDER BY b.amount ASC
        """),
        {"jid": job_id},
    )
    txn_result = await db.execute(
        text("""
            SELECT t.*, u.full_name as freelancer_name, u.profile_picture as freelancer_picture,
                   u.rating as freelancer_rating
            FROM transactions t
            LEFT JOIN users u ON t.freelancer_id = u.id
            WHERE t.job_id = :jid
            ORDER BY t.created_at DESC
        """),
        {"jid": job_id},
    )
    deliveries_result = await db.execute(
        text("SELECT * FROM job_deliveries WHERE job_id = :jid ORDER BY created_at DESC"),
        {"jid": job_id},
    )

    return {
        "job": dict(job),
        "bids": [dict(b) for b in bids_result.mappings().all()],
        "transactions": [dict(t) for t in txn_result.mappings().all()],
        "deliveries": [dict(d) for d in deliveries_result.mappings().all()],
    }


@router.post("")
async def create_job(req: JobCreateRequest, user: dict = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    if user["role"] != "client":
        raise HTTPException(403, "Only clients can create jobs")
    if not req.title or not req.description or not req.budgetMin or not req.budgetMax or not req.category:
        raise HTTPException(400, "Title, description, budget range, and category are required")

    job_id = str(uuid.uuid4())
    await db.execute(
        text("""
            INSERT INTO jobs (id, client_id, title, description, budget_min, budget_max, category, deadline)
            VALUES (:id, :cid, :title, :desc, :bmin, :bmax, :cat, :deadline)
        """),
        {
            "id": job_id, "cid": user["id"], "title": req.title, "desc": req.description,
            "bmin": req.budgetMin, "bmax": req.budgetMax, "cat": req.category,
            "deadline": req.deadline,
        },
    )
    result = await db.execute(text("SELECT * FROM jobs WHERE id = :id"), {"id": job_id})
    return {"job": dict(result.mappings().first())}


@router.post("/{job_id}/bid")
async def place_bid(job_id: str, req: BidRequest, user: dict = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    if user["role"] != "freelancer":
        raise HTTPException(403, "Only freelancers can bid")

    job_result = await db.execute(
        text("SELECT * FROM jobs WHERE id = :id AND status = 'open'"),
        {"id": job_id},
    )
    job = job_result.mappings().first()
    if not job:
        raise HTTPException(404, "Job not found or not accepting bids")

    if not req.amount:
        raise HTTPException(400, "Bid amount is required")

    existing = await db.execute(
        text("SELECT id FROM bids WHERE job_id = :jid AND freelancer_id = :fid"),
        {"jid": job_id, "fid": user["id"]},
    )
    if existing.mappings().first():
        raise HTTPException(409, "You already placed a bid on this job")

    bid_id = str(uuid.uuid4())
    await db.execute(
        text("INSERT INTO bids (id, job_id, freelancer_id, amount, proposal) VALUES (:id, :jid, :fid, :amt, :prop)"),
        {"id": bid_id, "jid": job_id, "fid": user["id"], "amt": req.amount, "prop": req.proposal},
    )

    # Notify client
    await db.execute(
        text("INSERT INTO notifications (id, user_id, title, message, type) VALUES (:id, :uid, :title, :msg, :type)"),
        {
            "id": str(uuid.uuid4()), "uid": job["client_id"],
            "title": "New Bid Received",
            "msg": f"{user['full_name']} placed a bid of ETB {req.amount} on your job \"{job['title']}\"",
            "type": "bid",
        },
    )

    bid_result = await db.execute(text("SELECT * FROM bids WHERE id = :id"), {"id": bid_id})
    return {"bid": dict(bid_result.mappings().first())}


@router.put("/{job_id}/award")
async def award_job(job_id: str, req: AwardRequest, user: dict = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    if user["role"] != "client":
        raise HTTPException(403, "Only clients can award jobs")

    job_result = await db.execute(
        text("SELECT * FROM jobs WHERE id = :id AND client_id = :cid AND status = 'open'"),
        {"id": job_id, "cid": user["id"]},
    )
    job = job_result.mappings().first()
    if not job:
        raise HTTPException(404, "Job not found or not open")

    bid_result = await db.execute(
        text("SELECT * FROM bids WHERE job_id = :jid AND freelancer_id = :fid AND status = 'pending'"),
        {"jid": job_id, "fid": req.freelancerId},
    )
    bid = bid_result.mappings().first()
    if not bid:
        raise HTTPException(404, "Bid not found")

    await db.execute(text("UPDATE jobs SET status = 'in_progress', awarded_to = :fid, updated_at = CURRENT_TIMESTAMP WHERE id = :id"), {"fid": req.freelancerId, "id": job_id})
    await db.execute(text("UPDATE bids SET status = 'accepted' WHERE job_id = :jid AND freelancer_id = :fid"), {"jid": job_id, "fid": req.freelancerId})
    await db.execute(text("UPDATE bids SET status = 'rejected' WHERE job_id = :jid AND freelancer_id != :fid AND status = 'pending'"), {"jid": job_id, "fid": req.freelancerId})

    txn_id = str(uuid.uuid4())
    await db.execute(
        text("INSERT INTO transactions (id, job_id, client_id, freelancer_id, amount, status) VALUES (:id, :jid, :cid, :fid, :amt, 'escrow')"),
        {"id": txn_id, "jid": job_id, "cid": user["id"], "fid": req.freelancerId, "amt": bid["amount"]},
    )

    await db.execute(
        text("INSERT INTO notifications (id, user_id, title, message, type) VALUES (:id, :uid, :title, :msg, :type)"),
        {
            "id": str(uuid.uuid4()), "uid": req.freelancerId,
            "title": "Job Awarded!",
            "msg": f"You have been awarded the job \"{job['title']}\"!",
            "type": "award",
        },
    )

    return {"message": "Job awarded successfully", "transactionId": txn_id}


@router.put("/{job_id}/deliver")
async def deliver_job(job_id: str, req: DeliverRequest, user: dict = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    job_result = await db.execute(
        text("SELECT * FROM jobs WHERE id = :id AND status = 'in_progress'"),
        {"id": job_id},
    )
    job = job_result.mappings().first()
    if not job:
        raise HTTPException(404, "Job not found or not in progress")
    if job["awarded_to"] != user["id"]:
        raise HTTPException(403, "Only the awarded freelancer can deliver")

    if req.files and len(req.files) > 0:
        if len(req.files) > 10:
            raise HTTPException(400, "Maximum 10 files per delivery")
        delivery_id = str(uuid.uuid4())
        await db.execute(
            text("INSERT INTO job_deliveries (id, job_id, freelancer_id, message, files) VALUES (:id, :jid, :fid, :msg, :files)"),
            {"id": delivery_id, "jid": job_id, "fid": user["id"], "msg": req.message or "", "files": json.dumps(req.files)},
        )

    await db.execute(text("UPDATE jobs SET status = 'delivered', updated_at = CURRENT_TIMESTAMP WHERE id = :id"), {"id": job_id})

    await db.execute(
        text("INSERT INTO notifications (id, user_id, title, message, type) VALUES (:id, :uid, :title, :msg, :type)"),
        {
            "id": str(uuid.uuid4()), "uid": job["client_id"],
            "title": "Work Delivered!",
            "msg": f"{user['full_name'] or 'The freelancer'} has delivered the finished work for \"{job['title']}\".",
            "type": "order",
        },
    )

    return {"message": "Work delivered successfully"}


@router.put("/{job_id}/status")
async def update_job_status(job_id: str, req: JobStatusUpdate, user: dict = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    job_result = await db.execute(text("SELECT * FROM jobs WHERE id = :id"), {"id": job_id})
    job = job_result.mappings().first()
    if not job:
        raise HTTPException(404, "Job not found")

    if req.status == "completed" and job["client_id"] != user["id"]:
        raise HTTPException(403, "Only the client can mark as completed")

    await db.execute(text("UPDATE jobs SET status = :s, updated_at = CURRENT_TIMESTAMP WHERE id = :id"), {"s": req.status, "id": job_id})
    return {"message": "Job status updated"}


@router.delete("/{job_id}")
async def delete_job(job_id: str, user: dict = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    job_result = await db.execute(
        text("SELECT * FROM jobs WHERE id = :id AND client_id = :cid"),
        {"id": job_id, "cid": user["id"]},
    )
    if not job_result.mappings().first():
        raise HTTPException(404, "Job not found or unauthorized")
    await db.execute(text("UPDATE jobs SET status = 'cancelled', updated_at = CURRENT_TIMESTAMP WHERE id = :id"), {"id": job_id})
    return {"message": "Job cancelled successfully"}


@router.post("/{job_id}/quick-order")
async def quick_order(job_id: str, req: QuickOrderRequest, user: dict = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    job_result = await db.execute(
        text("SELECT * FROM jobs WHERE id = :id AND status = 'open'"),
        {"id": job_id},
    )
    job = job_result.mappings().first()
    if not job:
        raise HTTPException(404, "Job not found or not open for orders")
    if not req.amount or req.amount <= 0:
        raise HTTPException(400, "A valid amount is required")

    await db.execute(
        text("UPDATE jobs SET status = 'in_progress', awarded_to = :fid, updated_at = CURRENT_TIMESTAMP WHERE id = :id"),
        {"fid": user["id"], "id": job_id},
    )

    txn_id = str(uuid.uuid4())
    await db.execute(
        text("INSERT INTO transactions (id, job_id, client_id, freelancer_id, amount, status) VALUES (:id, :jid, :cid, :fid, :amt, 'escrow')"),
        {"id": txn_id, "jid": job_id, "cid": job["client_id"], "fid": user["id"], "amt": req.amount},
    )

    await db.execute(
        text("INSERT INTO notifications (id, user_id, title, message, type) VALUES (:id, :uid, :title, :msg, :type)"),
        {
            "id": str(uuid.uuid4()), "uid": job["client_id"],
            "title": "New Order on Your Job!",
            "msg": f"{user['full_name'] or 'A freelancer'} placed a quick order on your job \"{job['title']}\" for ETB {req.amount:,.0f}.",
            "type": "order",
        },
    )

    return {"transactionId": txn_id, "message": "Quick order placed successfully!"}
